activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -1631,6 +1631,13 @@ def remove_timestamps_for_alter(table_name, **options)
           remove_columns_for_alter(table_name, :updated_at, :created_at)
         end
 
+        def add_foreign_key_for_alter(from_table, to_table, **options)
+          return unless supports_foreign_keys?
+
+          options = foreign_key_options(from_table, to_table, options)
+          schema_creation.send(:visit_AddForeignKey, ForeignKeyDefinition.new(from_table, to_table, options))
```

Comment:
I don't love the use of `send` here, but I see why it's necessary. It might possible to refactor the schema creation visitor to avoid it, but it's beyond the scope of this PR.

Reply:
Just use `schema_creation.accept`.

```suggestion
          schema_creation.accept ForeignKeyDefinition.new(from_table, to_table, options)
```

Reply:
I was originally going to suggest that too, but it doesn't work.

```ruby
$ ARCONN=mysql2 bin/test test/cases/migration_test.rb -n test_adding_multiple_columns
Using mysql2
Run options: -n test_adding_multiple_columns --seed 44489

# Running:

E

Error:
BulkAlterTableMigrationsTest#test_adding_multiple_columns:
ActiveRecord::StatementInvalid: Mysql2::Error: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'CONSTRAINT `fk_rails_f9f4f87832`
FOREIGN KEY (`parent_thing`)
  REFERENCES `dele' at line 1
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:206:in `block (2 levels) in execute'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:205:in `block in execute'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:726:in `block (2 levels) in log'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:725:in `block in log'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:716:in `log'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:204:in `execute'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb:48:in `execute'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1595:in `bulk_change_table'
    /Users/eugene/lib/rails/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:488:in `change_table'
    /Users/eugene/lib/rails/activerecord/test/cases/migration_test.rb:1305:in `with_bulk_change_table'
    /Users/eugene/lib/rails/activerecord/test/cases/migration_test.rb:1130:in `block in test_adding_multiple_columns'
    /Users/eugene/lib/rails/activerecord/test/cases/test_case.rb:54:in `assert_queries'
    /Users/eugene/lib/rails/activerecord/test/cases/migration_test.rb:1129:in `test_adding_multiple_columns'


bin/test test/cases/migration_test.rb:1120
```

```ruby
irb(#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00007fb3baab3108>):001:0> schema_creation.send(:visit_AddForeignKey, ForeignKeyDefinition.new(from_table, to_table, options))
=> "ADD CONSTRAINT `fk_rails_f9f4f87832`\nFOREIGN KEY (`parent_thing`)\n  REFERENCES `delete_me` (`id`)\n"
irb(#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00007fb3baab3108>):002:0> schema_creation.accept ForeignKeyDefinition.new(from_table, to_table, options)
=> "CONSTRAINT `fk_rails_f9f4f87832`\nFOREIGN KEY (`parent_thing`)\n  REFERENCES `delete_me` (`id`)\n"
```

`visit_ForeignKeyDefinition` is only useful when called internally by `visit_AddForeignKey`.

Reply:
Okey, so how about adding `AddForeignKeyDefinition` like #19989?
(indeed it might be beyond the scope of this PR as you said though)

https://github.com/rails/rails/blob/f109a3723887bfcf6b3fe43c59d0c4bea0ff8060/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L1599-L1603

```suggestion
          fk = ForeignKeyDefinition.new(from_table, to_table, options)
          schema_creation.accept AddForeignKeyDefinition.new(fk)
```

Reply:
Thanks @kamipo, that does look like a good approach.

@ccutrer if you're interested, feel free to explore this, otherwise I can follow up on it.

Reply:
updated PR introduces an intermediate object to avoid the send